src/snmpdump.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" snmpdump is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH SNMPDUMP 1 "September 07, 2006"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
snmpdump \- extract SNMP messages from pcap trace files
.SH SYNOPSIS
.B snmpdump
.RI [ options ]
.SH DESCRIPTION
\fBsnmpdump\fP extracts SNMP messages out of pcap trace files and
renders the messages in an XML or CSV format which can be easily processed
using scripts written in scripting languages such as Perl. Sensitive
data can be filtered or anonymized.
.SH OPTIONS
\fBsnmpdump\fP accepts the following options:
.TP
\fB-a, --anon\fP
Anonymize the trace. The anonymization rules are currently hardwired
and very restrictive and cannot be changed without recompiling the
program.
.TP
\fB-c \fIfile\fB, --config=\fIfile\fP
Read \fIfile\fP instead of any other (global and user)
libsmi configuration file.
.TP
\fB-f \fIexpression\fB, --filter=\fIexpression\fP
Filter the input by applying the pcap filter \fIexpression\fR to the
packets before they are processed by snmpdump. This option can only
be used once.
.TP
\fB-i \fIformat\fB, --input=\fIformat\fP
Process input of the given \fIformat\fP. The current version of
snmpdump can process XML input, PCAP input, and CSV input. The default
input format is PCAP.
.TP
\fB-o \fIformat\fB, --output=\fIformat\fP
Produce output of the given \fIformat\fP. The current version of
snmpdump can generate XML output and CSV output. The default output
format is XML.
.TP
\fB-w \fIfile\fB, --write=\fIfile\fP
Write output to \fIfile\fP instead of standard output.
.TP
\fB-p \fItext\fB, --passphrase=\fItext\fP
Provide a passphrase \fItext\fP as a parameter to the anonymization
transformations. This option has only effect if anonymization is
enabled. Note that specifying a passphrase on the command line is a
security risk since it might be visible in process lists.
.TP
.B \-t, \-\-translate
Translate version one traps into the format used by the second version
of the SNMP protocol operations. See RFC 3584 section 3.1. for
additional details.
.TP
\fB-m \fIfile\fB \-\-module \fIfile\fB
Load SMI MIB module definitions from \fIfile\fP.  This option can be
used several times to load several MIB modules into snmpdump.
.TP
\fB-z \fIregex\fB, --zap=\fIregex\fP
Clear all attributes or elements in the XML document whose name
matches \fIregex\fR. The regular expression \fIregex\fR is a case
insensitive extended POSIX regular expression.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-F, \-\-flows
Extract flows and write the output to several files which are named
after the flows. Any existing files will be overwritten. Messages that
cannot be assigned to flows will be written to standard output.
.TP
.B \-S, \-\-slice
Extract slices and write the output to several files which are named
after the slices. Any existing files will be overwritten. Messages that
cannot be assigned to slices will be written to standard output.
.TP
\fB-C \fIpath\fB, --change-path=\fIpath\fP
Write new flow files in \fIpath\fP rather than the current directory.
This option is only meaningful in combination with the flow option.
.TP
\fB-P \fIprefix\fB, --prefix=\fIprefix\fP
Generate flow file names that begin with the prefix \fIprefix\fP.
This option is only meaningful in combination with the flow option.
.TP
.B \-V, \-\-version
Show version of program.
.SH FORMATS
Two different output formats are generated by snmpdump: The XML format
is relatively verbose but preserves all information. The CSV format is
a more compact format which only keeps the most essential information
and is simple to process. For the details of the two format, see the
Internet Draft <draft-irtf-nmrg-snmp-measure-00.txt>.
.PP
The input formats accepted by snmpdump are the PCAP format, the XML
format, and the CSV format mentioned above. Note that CVS format can
only represent a subset of the available information.
.SH EXAMPLES
The following command converts SNMP traces stored in the 
file 'trace.pcap' into XML format.
.PP 
.RS 
\f(CWsnmpdump trace.pcap\fP
.RE
.PP
To convert the trace into CSV format, one would use the following command.
.PP
.RS
\f(CWsnmpdump -o csv trace.pcap\fP
.RE
.PP
In order to extract flows in CSV format which should be writting into flow
files located in /tmp, one would use the following command.
.PP
.RS
\f(CWsnmpdump -o csv -F -C /tmp trace.pcap\fP
.RE
.PP 
To zero out all community strings, you can use the following command
line:
.PP
.RS
\f(CWsnmpdump -z 'community' trace.pcap\fP
.RE
.PP
To filter out all vlen and blen attributes, use the following command
line:
.PP
.RS
\f(CWsnmpdump -z 'vlen|blen' trace.pcap\fP
.RE
.PP
To filter out only certain SNMP packets from a given source using a
non-standard port number, you can use the pcap filtering mechanism:
.PP 
.RS 
\f(CWsnmpdump -f 'udp and port 1161 and host 10.20.30.40' trace.pcap\fP
.RE 
.PP
To generate flow files in CSV format that are stored in the 
directory 'flows' with the flow file prefix 't01', use the following
command:
.PP
.RS
\f(CWsnmpdump -o csv -F -C flows -P t01 trace.pcap > flows/t01-unknown.csv\fP
.RE
.PP
.SH SEE ALSO
tcpdump(1)
.SH AUTHOR
Juergen Schoenwaelder <j.schoenwaelder@jacobs-university.de>.


ChangeLog
2005-06-12 23:49  schoenw

	* Makefile.in, acinclude.m4, aclocal.m4, autogen.sh, config.guess,
	  config.sub, configure, depcomp, doc/Makefile.in, install-sh,
	  missing, mkinstalldirs, src/Makefile.in, test/Makefile.in: Added
	  autogen.sh and removed all autoconf related files from the
	  repository.

2005-06-07 19:57  schoenw

	* AUTHORS, COPYING, ChangeLog, INSTALL, Makefile.am, Makefile.in,
	  NEWS, README, acinclude.m4, aclocal.m4, autom4te.cache,
	  autom4te.cache/output.0, autom4te.cache/requests,
	  autom4te.cache/traces.0, config.guess, config.sub, configure,
	  configure.in, depcomp, doc, doc/Makefile.am, doc/Makefile.in,
	  doc/snmptrace.rnc, install-sh, missing, mkinstalldirs, src,
	  src/Makefile.am, src/Makefile.in, src/config.h.in, src/snmpdump.1,
	  src/snmpdump.c, test, test/Makefile.am, test/Makefile.in,
	  test/frags.pcap, test/misc.pcap, test/scli.pcap, test/traps.pcap:
	  import snmpdump sources

2005-06-07 19:51  root

	* .: snmpdump project trunk



src/snmpdump.c
/*
 * snmpdump.c --
 *
 * A utility to convert pcap capture files containing SNMP messages
 * into snmp trace files. To create these pcap files, you can use:
 *
 *    tcpdump -i <interface> -s 0 -w <filename> udp and port 161 or port 162
 *
 * To convert the SNMP messages to XML, use this command:
 *
 *    snmpdump <filename>
 *
 * This implementation generates XML output directly rather than using
 * and XML writer API in order to be fast and memory efficient.
 *
 * Copyright (c) 2006 Juergen Schoenwaelder
 *
 * $Id$
 */

#define _GNU_SOURCE

#include "config.h"
#include "snmp.h"
#include "anon.h"

#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <regex.h>
#include <smi.h>

const char *progname = "snmpdump";

typedef enum {
    INPUT_XML = 1,
    INPUT_PCAP = 2,
    INPUT_CSV = 3
} input_t;

typedef enum {
    OUTPUT_XML = 1,
    OUTPUT_CSV = 2
} output_t;

#define STATE_FLAG_V1V2	0x01

typedef struct {
    uint64_t cnt;
    snmp_filter_t *filter;
    void (*do_filter)(snmp_filter_t *filter, snmp_packet_t *pkt);
    void (*do_learn)(snmp_packet_t *pkt);
    void (*do_anon)(snmp_packet_t *pkt);
    void (*do_flow_init)(snmp_write_t *out);
    void (*do_flow_write)(snmp_write_t *out, snmp_packet_t *pkt);
    void (*do_flow_done)(snmp_write_t *out);
    snmp_write_t out;
    int flags;
} callback_state_t;


/*
 * The per message callback which does all the processing and
 * printing, controlled by the state argument. This function is called
 * with a NULL packet pointer once we are done processing all packets.
 */

static void
print(snmp_packet_t *pkt, void *user_data)
{
    callback_state_t *state = (callback_state_t *) user_data;

    if (! state) {
	return;
    }

    /* Cleanup by printing the proper closing text in case we have
     * dealt with all packets.
     */

    if (! pkt) {
	if (state->do_flow_done) {
	    state->do_flow_done(&state->out);
	    return;
	}
	if (state->cnt && state->out.write_end && state->out.stream) {
	    state->out.write_end(state->out.stream);
	}
	return;
    }

    /* First apply the filters. Then call the anonymization module. We
     * might have to call it twice for learning purposes.
     */
    
    if (state->filter && state->do_filter) {
	state->do_filter(state->filter, pkt);
    }

    /*
     * Check whether we have to first apply any conversion. If yes, we
     * filter a second time since we might now have to apply
     * additional filter rules.
     */

    if (state->flags & STATE_FLAG_V1V2) {
	snmp_pkt_v1tov2(pkt);
	if (state->filter && state->do_filter) {
	    state->do_filter(state->filter, pkt);
	}
    }

    if (state->do_learn) {
	state->do_learn(pkt);
    }

    if (state->do_anon) {
	state->do_anon(pkt);
    }

    /*
     * Call the flow handler if it is set and we are done.
     */

    if (state->do_flow_write) {
	state->do_flow_write(&state->out, pkt);
	if (state->flags & STATE_FLAG_V1V2) {
	    snmp_pkt_delete(pkt);
	}
	return;
    }

    /* Otherwise, check whether we have to generate a header and then
     * print the packet.
     */

    if (state->cnt == 0 && state->out.stream && state->out.write_new) {
	state->out.write_new(state->out.stream);
    }

    if (state->out.stream && state->out.write_pkt) {
	state->out.write_pkt(state->out.stream, pkt);
    }
    state->cnt++;

    if (state->flags & STATE_FLAG_V1V2) {
	snmp_pkt_delete(pkt);
    }
}


/*
 * The main function to parse arguments, initialize the libraries and
 * to fire off the libnids library using nids_run() for every input
 * file we process.
 */

int
main(int argc, char **argv)
{
    int i, c;
    char *expr = NULL, *path = NULL, *prefix = NULL;
    output_t output = OUTPUT_XML;
    input_t input = INPUT_PCAP;
    char *errmsg;
    anon_key_t *key = NULL;
    callback_state_t _state, *state = &_state;
    FILE *stream = stdout;

    smiInit(progname);

    memset(state, 0, sizeof(*state));

    key = anon_key_new();
    anon_key_set_random(key);

    while ((c = getopt(argc, argv, "FSVz:f:w:i:o:c:m:hap:tC:P:")) != -1) {
	switch (c) {
	case 'a':
	    state->do_anon = snmp_anon_apply;
	    break;
	case 'z':
	    state->filter = snmp_filter_new(optarg, &errmsg);
	    if (! state->filter) {
		fprintf(stderr, "%s: ignoring clear filter: %s\n",
			progname, errmsg);
		continue;
	    }
	    state->do_filter = snmp_filter_apply;
	    break;
	case 'w':
	    stream = fopen(optarg, "w");
	    if (! stream) {
		fprintf(stderr, "%s: failed to open file %s: %s\n",
			progname, optarg, strerror(errno));
		exit(1);
	    }
	    break;
	case 'i':
	    if (strcmp(optarg, "pcap") == 0) {
		input = INPUT_PCAP;
	    } else if (strcmp(optarg, "xml") == 0) {
		input = INPUT_XML;
	    } else if (strcmp(optarg, "csv") == 0) {
		input = INPUT_CSV;
	    } else {
		fprintf(stderr, "%s: ignoring input format: %s unknown\n",
			progname, optarg);
	    }
	    break;
	case 'o':
	    if (strcmp(optarg, "csv") == 0) {
		output = OUTPUT_CSV;
	    } else if (strcmp(optarg, "xml") == 0) {
		output = OUTPUT_XML;
	    } else {
		fprintf(stderr, "%s: ignoring output format: %s unknown\n",
			progname, optarg);
	    }
	    break;
	case 'C':
	    path = optarg;
	    break;
	case 'P':
	    prefix = optarg;
	    break;
	case 't':
	    state->flags |= STATE_FLAG_V1V2;
	    break;
	case 'p':
	    anon_key_set_passphase(key, optarg);
	    break;
	case 'f':
	    expr = optarg;
	    break;
	case 'c':
	    smiReadConfig(optarg, progname);
	    break;
	case 'm':
	    smiLoadModule(optarg);
	    break;
	case 'F':
	    state->do_flow_write = snmp_flow_write;
	    state->do_flow_done = snmp_flow_done;
	    break;
	case 'S':
	    state->do_flow_write = snmp_slice_write;
	    state->do_flow_done = snmp_slice_done;
	    break;
	case 'V':
	    printf("%s %s\n", progname, VERSION);
	    exit(0);
	case 'h':
	case '?':
	    printf("%s [-c config] [-m module] [-f filter] [-i format] [-o format] [-z regex] [-p passphrase] [-w file] [-h] [-V] [-F] [-S] [-C path] [-P prefix] [-a] file ... \n", progname);
	    exit(0);
	}
    }

    state->out.stream = stream;
    state->out.write_new = NULL;
    state->out.write_pkt = NULL;
    state->out.write_end = NULL;
    state->out.path = path;
    state->out.prefix = prefix;

    if (state->do_anon) {
	anon_init(key);
    }

    switch (output) {
    case OUTPUT_XML:
	state->out.write_new = snmp_xml_write_stream_new;
	state->out.write_pkt = snmp_xml_write_stream_pkt;
	state->out.write_end = snmp_xml_write_stream_end;
	state->out.ext = "xml";
	break;
    case OUTPUT_CSV:
	state->out.write_new = snmp_csv_write_stream_new;
	state->out.write_pkt = snmp_csv_write_stream_pkt;
	state->out.write_end = snmp_csv_write_stream_end;
	state->out.ext = "csv";
	break;
    default:
	fprintf(stderr, "%s: unknown output format - aborting...\n", progname);
	abort();
    }

    if (optind == argc) {
	switch (input) {
	case INPUT_XML:
	    snmp_xml_read_stream(stdin, print, state);
	    break;
	case INPUT_PCAP:
	    snmp_pcap_read_stream(stdin, expr, print, state);
	    break;
	case INPUT_CSV:
	    snmp_csv_read_stream(stdin, print, state);
	    break;
	}
    } else {
	for (i = optind; i < argc; i++) {
	    switch (input) {
	    case INPUT_XML:
		snmp_xml_read_file(argv[i], print, state);
		break;
	    case INPUT_PCAP:
		snmp_pcap_read_file(argv[i], expr, print, state);
		break;
	    case INPUT_CSV:
		snmp_csv_read_file(argv[i], print, state);
		break;
	    }
	}
    }
    print(NULL, state);

    if (state->do_anon) {
	anon_done();
    }

    if (state->filter) {
	snmp_filter_delete(state->filter);
    }

    if (key) {
	anon_key_delete(key);
    }

    return 0;
}
